dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/persistence/file/EncryptedFileHandler.kt

```diff
@@ -0,0 +1,260 @@
+            return false
+        }
+
+        if (append && separator == null) {
+            internalLogger.e(MISSING_SEPARATOR_MESSAGE)
+            return false
+        }
+
+        val encryptedData = encryption.encrypt(data)
+
+        if (data.isNotEmpty() && encryptedData.isEmpty()) {
+            devLogger.e(BAD_ENCRYPTION_RESULT_MESSAGE)
+            return false
+        }
+
+        return delegate.writeData(
+            file,
+            // Base64 produces bytes per US-ASCII encoding, while separator may be in UTF-8 encoding
+            // but this is fine, because UTF-8 is backward compatible with US-ASCII (char in
+            // US-ASCII encoding has the same byte value as char in UTF-8 encoding)
+            base64Encoder(encryptedData),
+            append,
+            separator
+        )
+    }
+
+    override fun readData(
+        file: File,
+        prefix: ByteArray?,
+        suffix: ByteArray?,
+        separator: ByteArray?
+    ): ByteArray {
+
+        if (separator != null && !checkSeparator(separator)) {
+            internalLogger.e(INVALID_SEPARATOR_MESSAGE)
+            return EMPTY_BYTE_ARRAY
+        }
+
+        val data = delegate.readData(file, prefix, suffix, separator)
+
+        if (prefix == null && suffix == null && separator == null) {
+            return decryptSingleItemData(data)
+        }
+
+        val rawData = removeSuffixAndPrefix(data, prefix, suffix)
+
+        return if (separator != null) {
+            val decrypted = decryptBatchData(rawData, separator)
+            if (decrypted.isEmpty()) {
+                assemble(EMPTY_BYTE_ARRAY, prefix, suffix)
+            } else {
+                assemble(decrypted, prefix, suffix, separator)
+            }
+        } else {
+            val decrypted = decryptSingleItemData(rawData)
+            assemble(decrypted, prefix, suffix)
+        }
+    }
+
+    override fun delete(target: File) = delegate.delete(target)
+
+    override fun moveFiles(srcDir: File, destDir: File) = delegate.moveFiles(srcDir, destDir)
+
+    // endregion
+
+    // region private
+
+    private fun decryptSingleItemData(data: ByteArray): ByteArray {
+        val decoded = safeDecodeBase64(data)
+        return if (decoded.isNotEmpty()) {
+            encryption.decrypt(decoded)
+        } else {
+            decoded
+        }
+    }
+
+    private fun decryptBatchData(data: ByteArray, separator: ByteArray): List<ByteArray> {
+        return data
+            .splitBy(separator)
+            .map {
+                decryptSingleItemData(it)
+            }
+            .filter { it.isNotEmpty() }
+    }
+
+    private fun removeSuffixAndPrefix(
+        data: ByteArray,
+        prefix: ByteArray?,
+        suffix: ByteArray?
+    ): ByteArray {
+        return if (prefix != null || suffix != null) {
+            val prefixSize = prefix?.size ?: 0
+            val suffixSize = suffix?.size ?: 0
+
+            if (data.size < prefixSize + suffixSize) {
```

Comment:
Well, strictly speaking data can consist only from prefix and suffix if underlying file handler read empty byte array and I would expect this to be logged somewhere else. Will check the overall flow deeper.

Reply:
Tricky thing. Let's keep it like that, because the purpose of this method is to remove prefix and suffix and not to judge on the payload (even if it is 0 bytes).